chore: update README (#2)
stampery-labs/witnet-vanity: README.md

>
> — Satoshi Nakamoto in an email to Hal Finney in 2009

Witnet vanity addresses are are inspired in the [BIP 0173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki), i.e. they follow the Bech32 format. They are derived based on `Secp256k1` public keys and the hash function `SHA256`.

## Bech32 addresses

The Bech32 address format is structured as:

 - Human readeable part (hrp): type of data, or anything else that is relevant to the reader (MUST contain 1 to 83 US-ASCII characters).
 - Separator: always "1" by specification.
 - Data part: only consists of alphanumeric characters excluding "1", "b", "i", and "o". The last six characters of the data part is a checksum of `hrp` and `data` parts. The checksum is used to detect errors that may be introduced in transmission or storage.

More information can be found in the [BIP 0173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki).

## Usage

The tool has been written in Rust and it can be executed by using `cargo`. Installation can be found [here](https://www.rust-lang.org/tools/install).

> It is encourage to build the CLI tool with the `--release` tag for a better performance:

Reviewer: ```suggestion
It is encouraged to build the CLI tool with the `--release` tag for better performance:
```

    <vanity-string>    Vanity prefix string to generate
```   

The CLI tool arguments are:

 - \<vanity-string\>: mandatory argument to define the the vanity address starting prefix.
 - `--hrp` \<string\>: optional argument to define the human readable part of the Bech32 address. By default is set to `wit`.
 - `--threads` \<number\>: optional argument to set how many CPUs will be used for generating the vanity address. By default is set to use all available CPUs.

> The CLI tool will show an estimation of the progress while finding a valid vanity address, showing the amount of addresses checked and an estimated precessing time.

Reviewer: ```suggestion
The CLI tool will show an estimation of the progress while finding a valid vanity address, showing the amount of addresses checked and an estimated processing time.
```

```bash
./target/release/witnet-vanity example

Searching vanity addresses with the prefix: wit1example (threads: 8)

⠒ [00:00:31] [>-------------------------------------------------------------------------------] 3.2M/34.36G estimated tries (ETA: 3d)

```

### Additional considerations

Please take into account that valid vanity addresses are found by generating random `Secp256k1` key pairs. Therefore the estimation is just orientative and finding a valid address may take longer (or shorter) than the estimated time.

> Take into consideration the average required time to find a valid vanity address grows exponentially to the number of characters of the `vanity string`.

Reviewer: ```suggestion
Take into consideration that the average required time to find a valid vanity address grows exponentially with the number of characters of the `vanity string`.
```

 - address in bech32 format

### Generating a testnet Witnet address (`twit1`)

In order to generate a testnet Witnet address the `--hrp` argument should be set the human readable part to `twit`:

```bash
$ ./target/release/witnet-vanity m00n --hrp twit

Searching vanity addresses with the prefix: twit1m00n (threads: 8)

  [00:00:06] [###########################################################>--------------------] address found! (ETA: 0s)

Vanity address found:
        SK bytes:       07ccbeccfce73c191e0f9c95034a575a676f311326751203bdbc110ad0db66e5
        Private key:    xprv1qpnxu2vcp2a47d6sv7h4lpygnujhydhnhxsnjghe67recaqcvv5e7qq8ejlvel888sv3uruuj5p55466vahnzyexw5fq80duzy9dpkmxu5hn79vx
        Address:        twit1m00n7jmwkg2w7rr669frmay3h6e2733zm0cppp

```

#### Reusing addresses between mainnet and testnet

> Witnet testnet Bech32 addresses can be reused for different human readible parts (`hrp`). However, the last 6 address characters will changed as they are a checksum of the `hrp` and the `data` itself. In other words, the same private key (e.g. in `xprv` format) can be used to derive public keys for which may have different `hrp` parts.

Reviewer: ```suggestion
Witnet testnet Bech32 addresses can be reused for different human readible parts (`hrp`). However, the last 6 address characters will change as they are a checksum of the `hrp` and the `data` itself. In other words, the same private key (e.g. in `xprv` format) can be used to derive public keys for which may have different `hrp` parts.
```

> This is the case of the Witnet mainnet and testnet. For example, imagine we would like to reuse the previous found vanity address for mainnet (`wit1`):

Reviewer: ```suggestion
This is the case of the Witnet mainnet and testnet. For example, imagine we would like to reuse the previously found vanity address for mainnet (`wit1`):
```

```bash
Private key:    xprv1qq9qjt4xft2zyty744t7nuyergtulqhs52xrpcmt22hs3wzuk6tz6qrv8gtl9exxhntzmg9kvxgym5zs376mhnmzvluvam0u8fsgfxnm95s96kwv

>         Address:        wit1m00nj3eyzl5prcluexusfrcaf80fds6hhu5rxd

Reviewer: ```suggestion
Address:        wit1m00nj3eyzl5prcluexusfrcaf80fds6hhu5rxd
```

```

After importing the private key (with `xprv` format) in the node, the derived public key for Witnet testnet from the logs:

```bash
$ ./witnet node server --master-key-import master.key

[2020-06-09T18:15:42Z INFO  witnet_node::actors::chain_manager::actor] PublicKeyHash: twit1m00nj3eyzl5prcluexusfrcaf80fds6hefa8xu
```

> It is shown that the same `xprv` produces the the same valid vanity address but they differ in the `checksum` of in their Bech32 format representation (last 6 characters):

Reviewer: ```suggestion
It is shown that the same `xprv` produces the same valid vanity address but they differ in the `checksum` in their Bech32 format representation (last 6 characters):
```
